File: mbfpp/code/ClojureExamples/src/mbfpp/functional/mfr/discount.clj
(ns mbfpp.functional.mfr.discount)

(defn calculate-discount [prices]
  (reduce +
          (map (fn [price] (* price 0.10)) 
               (filter (fn [price] (>= price 20.0)) prices))))

(defn calculate-discount-namedfn [prices]
  (letfn [(twenty-or-greater? [price] (>= price 20.0))
          (ten-percent [price] (* price 0.10))]
         (reduce + 0.0 (map ten-percent (filter twenty-or-greater? prices)))))
         
(def bill [20.0 4.5 50.0 15.75 30.0 3.50])

File: mbfpp/code/ScalaExamples/src/main/scala/com/mblinn/mbfpp/functional/mfr/Discount.scala
package com.mblinn.mbfpp.functional.mfr

object Discount {

  def calculateDiscount(prices : Seq[Double]) : Double = {
    prices filter(price => price >= 20.0) map 
    	(price => price * 0.10) reduce 
    		((total, price) => total + price)
  }
  
  def calculateDiscountNamedFn(prices : Seq[Double]) : Double = {
    def isGreaterThan20(price : Double) = price >= 20.0
    def tenPercent(price : Double) = price * 0.10
    def sumPrices(total: Double, price : Double) = total + price
    
    prices filter isGreaterThan20 map tenPercent reduce sumPrices
  }
}